src/Radicle/Internal/Input.hs

```diff
@@ -0,0 +1,63 @@
+{-# ANN module "HLint: ignore Use <$>" #-}
+
+module Radicle.Internal.Input
+    ( InputT
+    , Completions
+    , noCompletions
+    , wordCompletions
+    , getInputLine
+    , runInputT
+    , outputStrLn
+
+    , MonadException
+    )
+where
+
+-- Haskeline-style wrapper, since it doesn't *quite* provide the right API
+-- (completion functions cannot be dynamically modified).  It will also enable
```

Comment:
The trouble is that we use `LangT ... (InputT IO)`, but the completion functions run in the underlying monad of `InputT`, namely `IO`.  So we don't have access to the binding environment which is up in the `LangT` layer.  

We *could* get around this by putting an `IORef` somewhere and reading it before we get a line.  Where to put it though?  If we put it up in `LangT`, then we've coupled `Lang` to `IO` when it was previously pure, which makes us lose all sorts of nice properties.  If we put it beneath `InputT`, then we basically have this solution again.

Besides `IORef`, another solution would be to invert the order of the transformers, so we have `InputT (Lang IO)`.  This might work, but it seems like it would be a pretty big change and might involve a lot of gymnastics to get `InputT` where we need it without having to talk about it in a bunch of irrelevant places.   I admit I haven't thought it through very much, willing to explore it more if you think it's worth it.

Reply:
> We could get around this by putting an IORef somewhere and reading it before we get a line. Where to put it though? If we put it up in LangT, then we've coupled Lang to IO when it was previously pure, which makes us lose all sorts of nice properties. If we put it beneath InputT, then we basically have this solution again.

Do we need to put it anywhere? If you create and pass it in to the completion functions in `main`, the type will still be `InputT IO`, no?

Reply:
I believe we do need to put it somewhere, because it needs to be visible by `get-expression` so that it can be written to.

Reply:
a) (sort of a more general question I still don't entirely understand, but) why do we need `get-expression` - i.e., why does completion stuff need to be exposed at the language level, rather than just having the previous system? Is the idea that you want completion to make sense post eval-redefinition? If so, what's the story there, exactly? 
b) if you do want the `get-expression`, and need an extra argument `r`, why wouldn't switching all `InputT IO`s to `InputT (ReaderT r IO)` work? (Or in the type synonym ReplM, just adding a `MonadReader r`?)

Reply:
(a) There was another option of defining what is now `repl.rad` in Haskell, where we read one line at a time and update the state.  But I chose to honor the current "fire and forget" implementation, where we run the interpreter and let it handle the rest, which necessitates exposing completion in language.  Indeed it's more flexible with eval redefinition, or even repl redefinition?  But I guess really I could have done it either way, and I did it this way.  What is your objection, exactly? 
(b) It does work, and that is what I have done. I just gave `ReaderT r IO` the name `InputT`, to communicate the intention of "patching" haskeline to have the needed feature.

Reply:
Here's what I mean: https://github.com/oscoin/radicle/commit/fc84bfb9f9b33eaee7a373e961fea768fa9566b1

(Note that this is quite messy and hurried.)

Reply:
Yes, I understood what you meant.  Your changes are, for the most part, aesthetic.  What you call `StateT [Text]` I call `InputT`.  What you call `(get-line! env)` I call `(get-expression! (keys env))` (completion is exposed at the language level either way).  Seems like that's about it.  

I'm not vehemently opposed to inlining my `InputT` abstraction as you have done, though I do think it's a step backwards as far as maintainability.  The only reason we needed any change to our monad stack was because haskeline lacked a needed feature. I would have preferred to patch haskeline (and maybe I still will), I just wanted to get this patch rolling without getting a third party into the mix.  So I made a wrapper around haskeline that does what I need.  Does my reasoning make sense?

I also consider the wrapper an advantage in that it hides the elements of haskeline that we do *not* depend on, thus leading the way for instantiating the abstraction in other ways.  I like to wrap neat little bows around components like this.